JIT: fast path for minopts/debug codegen in lvaMarkRefs

AndyAyersMS opened the pull request:
For minopts and debug codegen, consider all locals to be implicitly
referenced.

This is set up during `lvaMarkLocalVars` and maintained after that
by having `incRefCnts` set the implicit reference bit and not doing
anything in `decRefCnts` for minopts / debug.

Likewise suppress local var sorting, as we don't have accurate counts
to go by.

AndyAyersMS commented:
WIP for a couple of reasons:
* code size impact (~ 0.2%) bigger than I'd like for minopts. Main culprit is spilling unused register params to the stack. I've played with doing ref count scans for small methods (say 6 bytes of IL or less) and this cures some of it but doesn't really change TP any
* feels like like there need to be more safeguards after `lvaMarkRefs` is done so we don't inadvertently decrement ref counts to zero in minopts ... not sure the best way to code this just yet.

TP impact (via local PMI) is something around 5-7 % faster for minopts. Am running lab scenarios now to see if the numbers there agree.

Minimal diffs for non-minopts (just diffs in a handful of methods that force debug codegen, or cctors where we drop to minopts).

@dotnet-bot test Ubuntu x64 Checked minopts
@dotnet-bot test Windows_NT x64 Checked corefx_minopts
@dotnet-bot test Windows_NT x64 Checked minopts
@dotnet-bot test Windows_NT x86 Checked minopts

AndyAyersMS commented:
Looks like one bit of the logic in `lvaMarkLclRefs` is needed for minopts -- there's a part that sets the type of a variable. Seems likely we can just get the type right sometime earlier.

mikedn commented:
As far as I can tell this disables sorting in mark lclvars but we still sort for liveness before LSRA (in minopts mode). Is that needed?

AndyAyersMS commented:
There's a second part to this I haven't gotten to yet...

AndyAyersMS commented:
Lab numbers not quite so charitable, roughly a wash to 2% on tiering scenario jobs, and 1-2% on minopts crossgen. There is a somewhat more non-jit overhead in crossgen than there is in PMI so it's possible this 1-2% is is consistent with the 5%ish PMI results, but it seems a bit low.

I suppose I should run pure minopts scenarios.... or try measuring TP via SPMI which should have minimal non-jit overhead.

AndyAyersMS commented:
@dotnet-bot test Ubuntu x64 Checked minopts
@dotnet-bot test Windows_NT x64 Checked corefx_minopts
@dotnet-bot test Windows_NT x64 Checked minopts
@dotnet-bot test Windows_NT x86 Checked minopts

AndyAyersMS commented:
Not many gc stress legs for minopts, but here's one....

@dotnet-bot test Windows_NT x64 Checked gcstress0xc_minopts_heapverify1

AndyAyersMS commented:
Ubuntu minopts failure has an assert:
```
JIT/SIMD/VectorExp_ro/VectorExp_ro.sh

Assert failure(PID 79593 [0x000136e9], Thread: 79593 [0x136e9]): Assertion failed 'arg->OperGet() == GT_LCL_VAR' in 'VectorExpTest`1[Int64][System.Int64]:VectorExp(struct,long,long,long):int' (IL size 148)

src/jit/morph.cpp Line: 4630
```
GC stress failure possibly related:
```
JIT_HardwareIntrinsics._X86_Bmi1_Bmi1_r_Bmi1_r_._X86_Bmi1_Bmi1_r_Bmi1_r_cmd

Assert failure(PID 5276 [0x0000149c], Thread: 7592 [0x1da8]): !CREATE_CHECK_STRING(!\"Detected use of a corrupted OBJECTREF. Possible GC hole.\")
```
Other failures look iffy and possibly not related.

AndyAyersMS commented:
Morph assert happens w/o this change. We have this tree:
```
[000054] x---G+------                 /--*  OBJ(32)   simd32
[000053] ----G+------                 |  \--*  ADDR      byref
[000041] ----G+------                 |     |     /--*  LCL_VAR   simd32(AX) V06 loc2
[000044] ----G+------                 |     |  /--*  SIMD      simd32 long &~
[000043] -----+------                 |     |  |  \--*  LCL_VAR   simd32 V10 tmp2
[000045] ----G+-N----                 |     \--*  SIMD      simd32 long |
[000037] -----+------                 |        |  /--*  LCL_VAR   simd32 V10 tmp2
[000042] ----G+------                 |        \--*  SIMD      simd32 long &
[000039] ----G+------                 |           |  /--*  LCL_VAR   simd32(AX) V06 loc2
[000040] ----G+------                 |           \--*  SIMD      simd32 long -
[000038] -----+------                 |              \--*  LCL_VAR   simd32 V09 tmp1
[000046] -A--G+------              /--*  COMMA     simd32
[000026] ----G+------              |  |     /--*  LCL_VAR   simd32(AX) V06 loc2
[000033] ----G+-N----              |  |  /--*  SIMD      simd32 long gt
[000032] -----+------              |  |  |  \--*  LCL_VAR   simd32 V09 tmp1
[000036] -A--G+--R---              |  \--*  ASG       simd32 (copy)
[000034] D----+-N----              |     \--*  LCL_VAR   simd32 V10 tmp2
[000047] -A--G+------              *  COMMA     simd32
[000028] -----+-N----              |  /--*  SIMD      simd32 long init
[000027] -----+------              |  |  \--*  CNS_INT   long   0
[000031] -A---+--R---              \--*  ASG       simd32 (copy)
[000029] D----+-N----                 \--*  LCL_VAR   simd32 V09 tmp1
```
and `fgMorphMultiregStructArg` is not expecting a comma. Opened #19124.

AndyAyersMS commented:
BMI1 gcstress failure also repros without this change. Given this is a relatively new test I suspect it's likely a test bug, opened #19127.

AndyAyersMS commented:
Think using `lvImplicitlyReferenced` for all locals in minopts should give us the safeguards. So switching over to that now. Also will add bypassing some ref count updates in RCS_NORMAL for minopts/debug codegen.

AndyAyersMS commented:
Can't ignore increments all together as we're not resetting counts in minopts like Pat's change did. So if we see an increment in minopts/debug, make sure we note there is an implicit reference.

(implicit here in the sense that in minopts/debug we don't examine the IR for references, so all references are implicit... whereas in normal opts only the yet to be expanded, prolog/epilog and magic gc info references are implicit).

AndyAyersMS commented:
Redoing testing to see if we're past the recent versioning issue.

@dotnet-bot test this please

AndyAyersMS commented:
Lab data: crossgen TP shows 2-4% improvement. 

Scenario numbers for both minopts and tiering are again mostly a wash. There's a fair amount of noise in these but still you'd think startup metrics should show some wins.

Given crossgen and PMI data the minopts / debug / tier0 path is clearly somewhat faster. Why this doesn't seem to translate to wins on scenarios is a puzzle for another day.

AndyAyersMS commented:
Removed WIP comment, updated initial description. Think this is ready for review.

Throughput: 4-6% improvement in local PMI numbers for minopts; 2-4% improvement in lab crossgen numbers for minopts.

Code size: about 0.2 % increase in minopts, largely because of unreferenced args getting spilled. Will rerun and add data shortly.

@dotnet/jit-contrib PTAL

AndyAyersMS commented:
PMI minopts diffs (corelib partial as I'm still hitting #19140 where this is based locally)
```
Total bytes of diff: 86732 (0.20% of base)
    diff is a regression.

Top file regressions by size (bytes):
       16983 : Microsoft.CodeAnalysis.VisualBasic.dasm (0.27% of base)
       11360 : System.Private.CoreLib.dasm (0.25% of base)
       11223 : Microsoft.CodeAnalysis.CSharp.dasm (0.23% of base)
        6163 : System.Private.Xml.dasm (0.14% of base)
        4805 : System.Security.Permissions.dasm (15.13% of base)

Top file improvements by size (bytes):
        -643 : System.Reflection.Metadata.dasm (-0.14% of base)
        -494 : System.Linq.dasm (-0.06% of base)
        -317 : System.Numerics.Vectors.dasm (-0.34% of base)
        -277 : System.Drawing.Primitives.dasm (-0.58% of base)
        -265 : Microsoft.VisualBasic.dasm (-0.12% of base)

115 total files with size differences (22 improved, 93 regressed), 15 unchanged.

Top method regressions by size (bytes):
        1032 : Microsoft.CodeAnalysis.VisualBasic.dasm - _Closure$__:_Lambda$__0-0(ref):ref:this (258 methods)
         320 : System.Private.DataContractSerialization.dasm - XmlJsonWriter:WriteArray(ref,ref,ref,ref,int,int):this (20 methods)
         312 : xunit.runner.utility.dotnet.dasm - TestMessageVisitor:Visit(ref):bool:this (39 methods)
         296 : Microsoft.CodeAnalysis.VisualBasic.dasm - _Closure$__:_Lambda$__9-0(ref):ref:this (74 methods)
         276 : Microsoft.CodeAnalysis.VisualBasic.dasm - _Closure$__:_Lambda$__10-0(ref):ref:this (69 methods)

Top method improvements by size (bytes):
        -393 : System.Private.CoreLib.dasm - TaskFactory`1:ContinueWhenAll(ref,ref):ref:this (30 methods)
        -393 : System.Private.CoreLib.dasm - TaskFactory`1:ContinueWhenAny(ref,ref):ref:this (30 methods)
        -390 : System.Linq.Parallel.dasm - ConcatQueryOperator`1:WrapHelper(ref,ref,ref,struct,bool):this (25 methods)
        -390 : System.Private.CoreLib.dasm - Task`1:ContinueWith(ref,ref):ref:this (60 methods)
        -390 : System.Private.CoreLib.dasm - TaskFactory`1:ContinueWhenAll(ref,ref,int):ref:this (30 methods)

25158 total methods with size differences (3930 improved, 21228 regressed), 166152 unchanged.

14 files had text diffs but not size diffs.
```

Size increases come from spilling unused args. Size decreases from shorter prolog zeroing sequences (individual moves instead of rep mov sequence).

Experiments to allow ref counting in "small" methods for minopts / debug were successful at paring back some of the size increase, but didn't help TP any. So keeping minopts / debug behavior consistent seems like the way to go.

AndyAyersMS commented:
Local minopts PMI numbers for a few large assemblies

Assembly | Base | Diff | Ratio
--- | --- | --- | ---
System.Private.Xml | 1823 | 1728 | 1.055
Microsoft.CodeAnalysis.CSharp | 2774 | 2640 | 1.051

briansull commented on src/jit/lclvars.cpp:
```diff
@@ -3985,47 +4002,68 @@ void Compiler::lvaMarkLocalVars()
+    //
+    // Mark all locals as implicitly referenced and untracked.
+    // Don't bother sorting.
+    if (opts.MinOpts() || opts.compDbgCode)
```
We probably could use a helper method that checked these two.

AndyAyersMS replied:
Yeah, will put this on my todo list -- in particular there are places that just check for MinOpts which seems odd.

AndyAyersMS commented:
@dotnet-bot retest Ubuntu x64 Checked Innerloop Build and Test
